### PR TITLE
IPW Classical Mean

### PR DESCRIPTION
**Pull request by imerad**
### Description

- What does this PR do?
implements an estimator for Inverse probability weighted mean estimation
- Which issue does it close? (use `Closes #<number>`)
Fulfills this [ticket](https://github.com/orgs/EmertonData/projects/26/views/2?pane=issue&itemId=171028192)
- Any noteworthy implementation decisions or trade-offs?
Delayed integration into ASI doc notebooks for the ASI refactorization PRs

### Type of change

Checkbox list:
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Refactoring (no behavior change)
- [ ] Documentation
- [ ] Repository hygiene

### Checklist

Quality gates that must be satisfied before requesting a review:

- [x] I have read `CONTRIBUTING.md`
- [x] `make lint` passes
- [x] `make type-check` passes
- [x] `make tests` passes
- [x] `make coverage` reports 100% coverage
- [x] New public API has numpy-style docstrings
- [x] New public API is inserted in the API reference section of the documentation
- [x] Docs build without warnings (`make doc`)
- [x] `CHANGELOG.md` updated if the change is user-facing

### LLM usage

Disclose if an LLM was used in writing this PR:
- [ ] No LLM used
- [x] LLM used: ___
- [x] I went through and validated all the code myself

### CODE REVIEW

**Comment by imerad** [comment]
### Code review

Found 3 issues:

1. `_compute_ipw_weighted_values` uses `np.nan_to_num(y, nan=0)` which converts unsampled NaN entries to `0.0` before dividing by `sampling_probability`. Those zeros are then treated as real observations by `np.nanstd` in `_compute_std_estimate`, inflating the variance and producing an incorrect standard error (and therefore wrong confidence intervals) whenever any inputs are NaN. The fix is to preserve NaN for unobserved entries (e.g. `np.where(np.isnan(y), np.nan, y / sampling_probability)`) so that `nanstd` skips them.

https://github.com/EmertonData/glide/blob/831de51a7cc5c80e48ae312ff952f138b6077b51/glide/estimators/ipw_classical.py#L35-L45

2. The zero-variance guard checks uniqueness of the raw `y` values, but the relevant variance is that of the IPW-weighted values `y_i / π_i`. If all observed `y_i` are equal but their `π_i` differ, the weighted values have non-zero variance and are a valid input — yet the guard raises `ValueError` incorrectly. The check should be applied to `ipw_weighted_values`, not to `y`.

https://github.com/EmertonData/glide/blob/831de51a7cc5c80e48ae312ff952f138b6077b51/glide/estimators/ipw_classical.py#L75-L80

3. The docstring declares `π_i ∈ (0, 1]` but only the lower bound is validated (`np.min(sampling_probability) <= 0`). There is no check that `np.max(sampling_probability) <= 1`. A probability greater than 1 silently passes validation and produces statistically invalid IPW weights.

https://github.com/EmertonData/glide/blob/831de51a7cc5c80e48ae312ff952f138b6077b51/glide/estimators/ipw_classical.py#L60-L76

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

File: glide/estimators/ipw_classical.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    Point Estimate: 6.528
+    Confidence Interval (95%): [-1.230, 14.286]
+    Estimator : IPWClassicalMeanEstimator
+    n: 3
+    """
+
+    def _compute_ipw_weighted_values(self, y: NDArray, sampling_probability: NDArray) -> NDArray:
+        ipw_weighted_values = np.nan_to_num(y, nan=0) / sampling_probability
+        if len(np.unique(ipw_weighted_values)) == 1:
+            raise ValueError("Input values have zero variance")
```
Comment (mraki-ed) [inline glide/estimators/ipw_classical.py]: je me demande si c'est pas mieux de faire le if sur var == 0? pour traiter des edge case type on a 2 échantillons de valeurs 1 et 1.000001 le test passera mais la variance pourrait être trop proche de zéro en floating point?

File: glide/estimators/ipw_classical.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    >>> estimator = IPWClassicalMeanEstimator()
+    >>> result = estimator.estimate(y, pi)
+    >>> print(result)
+    Metric: Metric
+    Point Estimate: 6.528
+    Confidence Interval (95%): [-1.230, 14.286]
+    Estimator : IPWClassicalMeanEstimator
+    n: 3
+    """
+
```
Comment (mraki-ed) [inline glide/estimators/ipw_classical.py:34]: pas de `_preprocess`pour ipwclassical?

**Review by mraki-ed — DISMISSED** [review]

File: glide/estimators/ipw_classical.py (hunk 1 above)
Comment (imerad) [reply glide/estimators/ipw_classical.py]: Certes, mais je me demande si ça gère vraiment cet edge case car dans ce cas, la variance n'est pas garantie d'être nulle. Il faudrait aussi qu'on s'assure que cet aspect est géré de la même manière ailleurs dans le package. Hors je trouve la condition avec unique ailleurs. Je préfère créer un ticket dédié à ça

Comment (imerad) [reply glide/estimators/ipw_classical.py]: Voici le ticket https://github.com/orgs/EmertonData/projects/26?pane=issue&itemId=174537552

File: glide/estimators/ipw_classical.py (hunk 2 above)
Comment (imerad) [reply glide/estimators/ipw_classical.py:34]: Pas besoin en principe, je ne vois pas de prétraitement à faire

File: glide/estimators/ipw_classical.py (hunk 1 above)
Comment (mraki-ed) [reply glide/estimators/ipw_classical.py]: Oui effectivement on l'avait traité comme ça auparavant, à voir plus tard

File: glide/estimators/ipw_classical.py (hunk 2 above)
Comment (mraki-ed) [reply glide/estimators/ipw_classical.py:34]: D'accord

File: glide/estimators/ipw_classical.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        Raises
+        ------
+        ValueError
+            If any value in ``sampling_probability`` is not in ``(0, 1]``.
+        ValueError
+            If the IPW-weighted values have zero variance (all identical after
+            reweighting), which makes standard error estimation undefined.
+        """
+        if np.min(sampling_probability) <= 0 or np.max(sampling_probability) > 1:
+            raise ValueError(f"Minimum sampling probability should be in (0, 1], got {np.min(sampling_probability)}")
```
Comment (gmartinon-ed) [inline glide/estimators/ipw_classical.py]: Incohérence entre le if et le message.
Soit on fait deux raise différent (<0 et >1), soit on a un message qui donne le range de pi.
-> A propager dans ASI :-)

File: glide/estimators/ipw_classical.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    Point Estimate: 6.528
+    Confidence Interval (95%): [-1.230, 14.286]
+    Estimator : IPWClassicalMeanEstimator
+    n: 3
+    """
+
+    def _compute_ipw_weighted_values(self, y: NDArray, sampling_probability: NDArray) -> NDArray:
+        ipw_weighted_values = np.nan_to_num(y, nan=0) / sampling_probability
+        if len(np.unique(ipw_weighted_values)) == 1:
+            raise ValueError("Input values have zero variance")
```
Comment (gmartinon-ed) [inline glide/estimators/ipw_classical.py]: ça on peut le toper dans estimate non ?

File: glide/estimators/ipw_classical.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        return ipw_weighted_values
+
+    def _compute_mean_estimate(self, ipw_weighted_values: NDArray) -> float:
+        mean = np.mean(ipw_weighted_values)
```
Comment (gmartinon-ed) [inline glide/estimators/ipw_classical.py:40]: c'est pas nanmean ici ? En cohérence avec le ClassicalMeanEstimator ?

File: glide/estimators/ipw_classical.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        return mean
+
+    def _compute_std_estimate(self, ipw_weighted_values: NDArray) -> float:
+        std = np.std(ipw_weighted_values, ddof=1) / np.sqrt(len(ipw_weighted_values))
```
Comment (gmartinon-ed) [inline glide/estimators/ipw_classical.py:44]: C'est pas nan_std ici, en cohérence avec le ClassicalMeanEstimator ?
Pareil pour le dénominateur, pour moi c'est la taille en excluant les NaN

File: tests/functional/estimators/test_ipw_classical.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+    labeled_dataset, _ = generate_gaussian_dataset(n, 0, true_mean=true_mean, true_std=true_std, random_seed=0)
+    y = labeled_dataset["y_true"]
+    sampling_probability = np.clip(rng.random(n), 0.5, 1)
+    y[rng.random(n) > sampling_probability] = np.nan
+
+    result = IPWClassicalMeanEstimator().estimate(y, sampling_probability)
+
+    assert result.mean == pytest.approx(true_mean, abs=0.2)
+    assert result.std == pytest.approx(0.11, abs=0.01)
```
Comment (gmartinon-ed) [inline tests/functional/estimators/test_ipw_classical.py:38]: pourquoi pas `true_std` ?

File: tests/functional/estimators/test_ipw_classical.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    assert result.n == np.sum(~np.isnan(y))
+
+
+def test_uniform_sampling_probability_matches_classical():
+    """IPW with all sampling probabilities equal to 1 on fully observed data reduces to classical mean estimator.
```
Comment (gmartinon-ed) [inline tests/functional/estimators/test_ipw_classical.py]: Et quand il y a des valeurs manquantes dans y ? (cf mes commentaires sur la classe elle-même)
idéalement, on voudrait pouvoir appliquer le IPWClassical sur les données natives de GLIDE, donc avec des valeurs manquantes (cf les tutoriels). Les valeurs manquantes ne doivent rien changer à l'estimation pour moi.

-> ça peut faire un bon test fonctionnel ça d'ailleurs, à répliquer sur ClassicalMeanEstimator d'ailleurs

File: glide/estimators/ipw_classical.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        return std
```
Comment (gmartinon-ed) [inline glide/estimators/ipw_classical.py:45]: Tiens au passage, je remarque qu'il y a des `float()` en trop dans asi.py. possible de rattraper le coup ?

File: tests/unit/estimators/test_ipw_classical.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+def test_estimate_custom_confidence_level(estimator, y, sampling_probability):
+    result = estimator.estimate(y, sampling_probability, confidence_level=0.85)
+    assert result.confidence_interval.confidence_level == 0.85
+    assert result.confidence_interval.mean == pytest.approx(3.0)
+    assert result.std == pytest.approx(1.915, abs=0.001)
+    assert result.confidence_interval.lower_bound == pytest.approx(0.243, abs=0.001)
+    assert result.confidence_interval.upper_bound == pytest.approx(5.756, abs=0.001)
+
+
+@pytest.mark.parametrize("bad_pi", [2.0, 0.0, -0.5])
```
Comment (gmartinon-ed) [inline tests/unit/estimators/test_ipw_classical.py]: Réduire à 2 (<= 0 ou > 1)

File: tests/unit/estimators/test_ipw_classical.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    pi = np.array([0.5, 0.5, 0.5, bad_pi])
+    with pytest.raises(ValueError, match="Minimum sampling probability should be in \\(0, 1]"):
+        estimator.estimate(y, pi)
+
+
+def test_estimate_raises_on_zero_variance(estimator):
+    y = np.array([3.0, 3.0, 3.0])
+    pi = np.array([0.5, 0.5, 0.5])
+    with pytest.raises(ValueError, match="Input values have zero variance"):
+        estimator.estimate(y, pi)
```
Comment (gmartinon-ed) [inline tests/unit/estimators/test_ipw_classical.py]: Est-ce que c'est grave d'avoir une variance de 0 en classique ? J'ai l'impression que c'est OK, car contrairement à PPI etc., il n'y a jamais de division par la variance.

Je retirerais ce test, et le raise qui va avec dans la classe

**Review by gmartinon-ed — CHANGES_REQUESTED** [review]

File: glide/estimators/ipw_classical.py (hunk 3 above)
Comment (imerad) [reply glide/estimators/ipw_classical.py]: Traité

File: glide/estimators/ipw_classical.py (hunk 4 above)
Comment (imerad) [reply glide/estimators/ipw_classical.py]: Fait

File: glide/estimators/ipw_classical.py (hunk 5 above)
Comment (imerad) [reply glide/estimators/ipw_classical.py:40]: Non, ici les NaNs correspondent à des instances non observées et sont remplacés par des zeros avant le calcul (correspondant à un xi = 0), j'ai préféré garder un np.mean car on n'est pas censés ignorer les NaNs ici

File: glide/estimators/ipw_classical.py (hunk 6 above)
Comment (imerad) [reply glide/estimators/ipw_classical.py:44]: Idem que commentaire précédent. Les NaNs doivent être traités comme des zeros ici. On doit normaliser par rapport au nombre total d'instances, pas le nombre d'instances observées

File: tests/functional/estimators/test_ipw_classical.py (hunk 7 above)
Comment (imerad) [reply tests/functional/estimators/test_ipw_classical.py:38]: La variance de l'estimateur utilisé ici ne sera généralement pas la même que la variance de la variable échantillonnée. On a $\E[\xi Y / \pi] = \E [Y]$ mais pas $$\E[(\xi Y / \pi)^2] = \E [Y^2]$$, l'espérance sera généralement différente pour moment d'ordre 2, donc la variance aussi

File: tests/functional/estimators/test_ipw_classical.py (hunk 8 above)
Comment (imerad) [reply tests/functional/estimators/test_ipw_classical.py]: Contrairement à Classical, le IPWClassical ne doit pas ignorer les NaNs, il les considère comme des zeros et normalise par la longueur totale de l'échantillon, cf mes autres commentaires.

Comment (imerad) [reply tests/functional/estimators/test_ipw_classical.py]: Je rajoute des tests unitaires vérifiant que les NaNs sont ignorés dans classical

File: glide/estimators/ipw_classical.py (hunk 9 above)
Comment (imerad) [reply glide/estimators/ipw_classical.py:45]: Fait

File: tests/unit/estimators/test_ipw_classical.py (hunk 10 above)
Comment (imerad) [reply tests/unit/estimators/test_ipw_classical.py]: Fait

File: tests/unit/estimators/test_ipw_classical.py (hunk 11 above)
Comment (imerad) [reply tests/unit/estimators/test_ipw_classical.py]: Ça donne un intervalle de confiance trivial .. Mais ok, c'est enlevé

File: glide/estimators/ipw_classical.py (hunk 4 above)
Comment (gmartinon-ed) [reply glide/estimators/ipw_classical.py]: Ah bon ?

File: glide/estimators/ipw_classical.py (hunk 5 above)
Comment (gmartinon-ed) [reply glide/estimators/ipw_classical.py:40]: Yes, mais sachant qu'on a un ticket de refacto qui vise justement à mettre cette fonction dans un `core` consommé à la fois par IPWClassial et Classical, il n'y a pas moyen d'homogénéiser en anticipation ?

File: glide/estimators/ipw_classical.py (hunk 4 above)
Comment (imerad) [reply glide/estimators/ipw_classical.py]: Je l'avais mis dans `estimate` mais finalement retiré car un autre commentaire proposait d'enlever le test de  variance nulle

File: glide/estimators/ipw_classical.py (hunk 5 above)
Comment (imerad) [reply glide/estimators/ipw_classical.py:40]: Une manière simple de le faire serait de demander à l'utilisateur de mettre des zeros pour les items non-observés pour ne plus devoir le faire dans `_compute_ipw_weighted_values`. À ce moment, on pourra ignorer les NaNs. Qu'en penses tu ?

**Review by gmartinon-ed — APPROVED** [review]
Ok, je suis convaincu par ces dénominateurs dans mean et std. Il y aura un sujet de factorisation dans le sprint N+1 pour savoir ce qu'on mets dans core/classical.py. Mais ce sera traité au prochain sprint
